elastic/apm-server
# Prevent index failed precondition

this patch closes the `publishReady` channel only
if all preconditions have been met. Previously,
any result from the waitReady check would result
in unblocking the roundtripper that was supposed
to prevent indexing until all preconditions had
been met.

when the server is shut down, the ctx in
s.waitReady is canceled. this caused the function
to return and execute its defer statement, closing
`publishReady`.  this resulted in any queued events
being sent to the elasticsearch, which is not the
desired behavior.

closes https://github.com/elastic/apm-server/issues/8018

# how to test

1. start elasticsearch/kibana
2. navigate to fleet ui / agent policies; remove apm-1
3. go to add integrations / apm / settings; remove the integration
4. start apm server; verify preconditions aren't met from log output
5. send events to apm-server
6. stop apm-server
7. verify no events were indexed
